File: core/java/src/main/java/hara/lang/protocol/IMetadata.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.imetadata", name = "IMetadata")
public interface IMetadata {

  @HaraMethod(value = "metatype", arity = 1)
  Constant.MetaType getMetatype();
}

File: core/rust/src/lang/protocol/imetadata.rs
use hara_protocol_macros::hara_protocol;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum MetaType {
    Object,
    Map,
    String,
}

#[hara_protocol(
    namespace = "std.protocol.imetadata",
    name = "IMetadata",
    availability = "portable"
)]
pub trait IMetadata: Sized {
    type Metadata: Clone;

    fn meta(&self) -> Option<&Self::Metadata>;
    fn with_meta(&self, metadata: Option<Self::Metadata>) -> Self;

    #[hara_method(value = "metatype", arity = 1)]
    fn metatype(&self) -> MetaType {
        MetaType::Object
    }
}
